### ISSUE TITLE
switch to rustup instead of multirust

### ISSUE THREAD

**Issue by NobbZ** [hansihe/Rustler#14]
The required version of rustc is only a week old, but still you are trying to use a version manager that is deprecated in favor of rustup for about half a year

**Comment by hansihe** [comment]
As far as I can tell, the deprecation warning for multirust was added only five days ago, in [this](https://github.com/brson/multirust/commit/56e2ece7e68251f770ab3303d350f8de6a12819c) commit. Was there another warning somewhere else?

That being said, moving to rustup is probably a good idea.

**Comment by NobbZ** [comment]
I thought that I have read about the deprecation about half a year ago for the first time, but I just skimmed through the logs as well and concluded for my self, that my memory is tricking me.

But since I did not worked with rust for a long time, I did a `multirust update` today and then it desintegrated by removing `multirust` executable as well as the proxy-binaries, so I was forced to install `rustup`, and I did also already fire up another issue related to it (#15).

**Comment by hansihe** [comment]
Oh, I didn't know multirust self-destructed.

I will work on rustup support as soon as I can.

**Comment by NobbZ** [comment]
And I can't tell if it was related to the deprecation or just something on my system or even doing something wrong when updating, all I can tell is that now there seems to be no way back.

PS: Why even check for multirust/rustup or any version manager instead of simply checking for the version of `rustc`?

**Comment by hansihe** [comment]
That is a very good question.

The mix compiler actually runs `multirust run` directly. When originally writing the mix compiler, I needed to run the build in many different environments, including dev machine, travis, and server. Having the compiler work with multirust directly made this easier in some ways, as it enabled me to have tasks for automatically fetching the correct version, and more helpful error messages for the user when compilation fails.

Making the compile task call rustc instead of multirust/rustup would enable users to choose their own rustc version management strategy. However doing that would mean less rigid sanity checks, and less helpful error messages.

I am still undecided on what to do. Do you have any more thoughs on this?

**Comment by NobbZ** [comment]
Rely on rustup/multirust per default, but make it overridable.

So the default user will be guided, but when someone overrides rustlers sanity checks, then he is probably knowing what he is doing hand has to debug by himself when something breaks away.

But how exactly this overide should work, or which commands it should run I'm not quite sure, probably the same as the checked version does, but without prefixing `multirust run`?

**Comment by hansihe** [comment]
Good call, making a sensible default, and making it overridable for users who know what they are doing is probably a good way of doing it.

**Comment by NobbZ** [comment]
Has there been any progress on this?

**Comment by hansihe** [comment]
There has, I basically have it working locally.

Actually pushing it is blocked by an issue with rustup.rs, rust-lang-nursery/rustup.rs#497. If I where to push it before that gets fixed in a release, it would complicate testing for projects using the rustler compiler.

It seems like the issue may have been fixed in the git repo already, but I am unsure when the next release will be.

**Comment by NobbZ** [comment]
Can you push to a wipe branch until then? Perhaps I can start playing
around using that branch.

Hans Elias J. <notifications@github.com> schrieb am Sa., 18. Juni 2016
16:25:

**Comment by hansihe** [comment]
Pushed it to the rustup branch.

**Comment by hansihe** [comment]
A new version of rustup has been released, the problem blocking this seems to be fixed now.

Will update and release a new version as soon as possible.

**Comment by hansihe** [comment]
The newest versions now use rustup.
